protocol/messenger_handler.go

```diff
@@ -2762,55 +2767,90 @@ func (m *Messenger) updateUnviewedCounts(chat *Chat, mentionedOrReplied bool) {
 	}
 }
 
-func (m *Messenger) HandleSyncWalletAccount(state *ReceivedMessageState, message protobuf.SyncWalletAccounts) error {
-	dbAccounts, err := m.settings.GetAccounts()
-	if err != nil {
-		return err
+func (m *Messenger) handleSyncWalletAccount(message *protobuf.SyncWalletAccount) (*accounts.Account, error) {
+	if message.Chat {
+		return nil, ErrNotWalletAccount
 	}
-
-	dbAccountMap := make(map[types.Address]*accounts.Account)
-	for _, acc := range dbAccounts {
-		dbAccountMap[acc.Address] = acc
+	accAddress := types.BytesToAddress(message.Address)
+	dbAccount, err := m.settings.GetAccountByAddress(accAddress)
+	if err != nil && err != sql.ErrNoRows {
+		return nil, err
 	}
 
-	var accs []*accounts.Account
-	for _, message := range message.Accounts {
-		dbAcc := dbAccountMap[types.BytesToAddress(message.Address)]
-		if dbAcc != nil && message.Clock <= dbAcc.Clock {
-			continue
-		}
-		var acc *accounts.Account
-		if dbAcc != nil && message.Removed {
-			acc = &accounts.Account{
-				Address: types.BytesToAddress(message.Address),
-				Removed: true,
-			}
-		} else if !message.Removed {
-			acc = &accounts.Account{
-				Address:   types.BytesToAddress(message.Address),
-				Wallet:    message.Wallet,
-				Chat:      message.Chat,
-				Type:      accounts.AccountType(message.Type),
-				Storage:   message.Storage,
-				PublicKey: types.HexBytes(message.PublicKey),
-				Path:      message.Path,
-				Color:     message.Color,
-				Hidden:    message.Hidden,
-				Name:      message.Name,
-				Clock:     message.Clock,
-			}
-
-		}
-
-		if runtime.GOOS != "android" && runtime.GOOS != "ios" {
-			// For the desktop app we need to ignore accounts with empty `KeypairName` or empty `DerivedFrom` (except if an account
-			// is not private key imported account or watch only account). Otherwise keypair items will be broken.
+	// Once mobile app supports seed phrase and private key imported accounts we should remove the following `if` block
+	if isMobileApp {
+		accType := accounts.AccountType(message.Type)
+		if accType != accounts.AccountTypeWatch &&
+			accType != accounts.AccountTypeGenerated {
+			return nil, ErrWalletAccountNotSupportedForMobileApp
+		}
+	}
+
+	if dbAccount != nil && message.Clock <= dbAccount.Clock {
```

Comment:
Could you explain `message.Clock`? `message.Clock` is generated by m.getLastClockWithRelatedChat(), so looks like it's always greater than the clock in db, what if dbAccount.Clock get updated by user on device B, then we received the backup message from device A? assuming we have the same account on device A/B @saledjenic

Reply:
@qfrank if you look more detailed you will see that the code here, technically, was not changed at all. It's exactly the same as it was before, what I did actually is that I extracted one part from the old [HandleSyncWalletAccount](https://github.com/status-im/status-go/blob/1fe4a898b60694077c6df6fef352c71bb9b7d8eb/protocol/messenger_handler.go#L2765) function into `handleSyncWalletAccount` that I can reuse that function in `handleBackedUpWalletAccount` function. That's all.

> message.Clock is generated by m.getLastClockWithRelatedChat()

This is true.

> so looks like it's always greater than the clock in db

This is not entirely true. Cause there are 3 use cases for this part. 
1. `recover an account, applying previously backed up data` - since data were backed up in some point in the past after which user might update an account, in that case clock from the backed up message is older then the clock stored in the db.   Elaboration: since backed up messages are applied during onboarding (when user recovers an account) they should always be applied (cause at that moment the db is empty) but there is always a possibility that some messages come later, in meanwhile user may do some local changes and a message coming from the backup, related to the same account should be discarded in that case.
2. `recover an account - syncing with another device` - since this action is available during onboarding (when user recovers an account) and that only if devices are connected to the same local network then messages should always be applied (cause at that moment the db is empty)
3. `sync among paired devices` - in this case devices which are paired do not need to be close to each other, they communicate over network, so there is no guarantee that a change made on device A will be immediately visible on the device B, that means if user make a change on the device be in meantime after it has been made on the device A, when device B receives a change from the device A it should discard it, cause it has clock value less than value stored on the device B.

Hope that above clarifies possible confusion.

Reply:
thank you for detailed explaination ❤️ but i still have the same question as @cammellos [mentioned](https://github.com/status-im/status-go/pull/3416#discussion_r1173435288) , his comment might looks more clear:
> it might override an actual decision of the user on a different device.